### PR TITLE
[jit] Allow dicts in C++ frontend

### LINKED ISSUE #16856: Type cannot be accurately recovered from this IValue for unordered_map
## 🐛 Bug
Hi, I'm trying to pass a dictionary to a jit model. It fails when checking that the input type matches the schema, in the `c10::isSubvalueOf` or `c10::incompleteInferTypeFrom` method. 

The stack trace is as follows:
```
Terminate c10::Error(Type cannot be accurately recovered from this IValue. (incompleteInferTypeFrom at ../aten/src/ATen/core/type.cpp:149) frame #0: c10::Error::Error(c10::SourceLocation, std::string const&) + 0x45 ( in ../lib/third_party/libc10.so) frame #1: c10::incompleteInferTypeFrom(c10::IValue const&) + 0x3b4 ( in ../lib/third_party/libcaffe2.so) frame #2: c10::isSubvalueOf(c10::IValue const&, std::shared_ptr<c10::Type>) + 0x25 ( in ../lib/third_party/libcaffe2.so) frame #3: <unknown function> + 0x5b54d0 ( in ../lib/third_party/libcaffe2.so) frame #4: c10::isSubvalueOf(c10::IValue const&, std::shared_ptr<c10::Type>) + 0x15e ( in ../lib/third_party/libcaffe2.so) frame #5: torch::jit::script::Method::checkInputsAgainstSchema(std::vector<c10::IValue, std::allocator<c10::IValue> >&) + 0x268 ( in ../lib/libml_libtorch_estimator.so) frame #6: torch::jit::script::Module::forward(std::vector<c10::IValue, std::allocator<c10::IValue> >) + 0x71 ( in ../lib/libml_libtorch_estimator.so) frame #7: std::map<std::string, at::Tensor, std::less<std::string>, std::allocator<std::pair<std::string const, at::Tensor> > > ml::libtorch_estimator::infer<at::Tensor>(bool) + 0x228 ( in ../lib/libml_libtorch_estimator.so) frame #8: load_and_run_model(std::string) + 0x1643 ( in ./bin/inference) frame #9: main + 0x154 ( in ./bin/inference) frame #10: __libc_start_main + 0xf5 ( in /lib/x86_64-linux-gnu/libc.so.6) frame #11: ./bin/inference() [] ) {
  libstdc++.so.6.0.19 5e6d6 eh_terminate.cc:42 __cxxabiv1::__terminate(void (*)())
  libstdc++.so.6.0.19 5e703
  libstdc++.so.6.0.19 5e922
  libcaffe2.so 5b4dbb c10::incompleteInferTypeFrom(c10::IValue const&)
  libcaffe2.so 5b5275 c10::isSubvalueOf(c10::IValue const&, std::shared_ptr<c10::Type>)
  libcaffe2.so 5b54d0 c10::isSubvalueOf(c10::IValue const&, std::shared_ptr<c10::Type>)::{lambda(c10::IValue const&)#1}::operator()(c10::IValue const&) const
  libcaffe2.so 5b53ae c10::isSubvalueOf(c10::IValue const&, std::shared_ptr<c10::Type>)
  libml_libtorch_estimator.so 1c448 torch::jit::script::Method::checkInputsAgainstSchema(std::vector<c10::IValue, std::allocator<c10::IValue> >&)
  libml_libtorch_estimator.so 141d1 torch::jit::script::Module::forward(std::vector<c10::IValue, std::allocator<c10::IValue> >)
  libml_libtorch_estimator.so 10b48 std::map<std::string, at::Tensor, std::less<std::string>, std::allocator<std::pair<std::string const, at::Tensor> > > ml::libtorch_estimator::infer<at::Tensor>(bool)
  inference 40c483 load_and_run_model(std::string)
  inference 40e184 main
  libc-2.19.so 21f45 libc-start.c:321 __libc_start_main
  inference 40ad77 _start
}
terminate called after throwing an instance of 'c10::Error'
  what():  Type cannot be accurately recovered from this IValue. (incompleteInferTypeFrom at ../aten/src/ATen/core/type.cpp:149)
frame #0: c10::Error::Error(c10::SourceLocation, std::string const&) + 0x45 ( in ../lib/third_party/libc10.so)
frame #1: c10::incompleteInferTypeFrom(c10::IValue const&) + 0x3b4 ( in ../lib/third_party/libcaffe2.so)
frame #2: c10::isSubvalueOf(c10::IValue const&, std::shared_ptr<c10::Type>) + 0x25 ( in ../lib/third_party/libcaffe2.so)
frame #3: <unknown function> + 0x5b54d0 ( in ../lib/third_party/libcaffe2.so)
frame #4: c10::isSubvalueOf(c10::IValue const&, std::shared_ptr<c10::Type>) + 0x15e ( in ../lib/third_party/libcaffe2.so)
frame #5: torch::jit::script::Method::checkInputsAgainstSchema(std::vector<c10::IValue, std::allocator<c10::IValue> >&) + 0x268 ( in ../lib/libml_libtorch_estimator.so)
frame #6: torch::jit::script::Module::forward(std::vector<c10::IValue, std::allocator<c10::IValue> >) + 0x71 ( in ../lib/libml_libtorch_estimator.so)
frame #7: std::map<std::string, at::Tensor, std::less<std::string>, std::allocator<std::pair<std::string const, at::Tensor> > > ml::libtorch_estimator::infer<at::Tensor>(bool) + 0x228 ( in ../lib/libml_libtorch_estimator.so)
frame #8: load_and_run_model(std::string) + 0x1643 ( in ./bin/inference)
frame #9: main + 0x154 ( in ./bin/inference)
frame #10: __libc_start_main + 0xf5 ( in /lib/x86_64-linux-gnu/libc.so.6)
frame #11: ./bin/inference() []
```

## To Reproduce

The python model's input is defined like so:
```
@torch.jit.script_method
    def forward(self, list_of_frame_dictionaries):
        # type: (List[Dict[str, Tensor]]) -> Tuple[Tensor, Tensor, Tensor, Tensor, Tensor] # noqa
```
The C++ code to format the data:
Method param: `const std::vector<std::unordered_map<std::string, torch::Tensor>> input_data`
```
    std::vector<c10::IValue> input;
    for (size_t pos = 0; pos < input_data.size(); ++pos)
    {
        c10::ivalue::DictUnorderedMap<c10::IValue, c10::IValue> ivalue_map;
        for (auto x: input_data[pos])
        {
            c10::IValue key = x.first;
            c10::IValue value = x.second.to(device_);
            ivalue_map[key] = value;
        }
        c10::IValue map_as_ivalue = ivalue_map;
        input.push_back(map_as_ivalue);
    }
    auto arg_index = schema.argumentIndexWithName(input_name);
    std::vector<torch::jit::IValue> inputs_;
    inputs_.at(arg_index.value()) = input;
```

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

PyTorch version: 1.1.0a0+a3f600e_1_ge44ddff
Python version: 2.7

### PR DESCRIPTION
**Pull request by driazati**
Fixes #16856

### CODE REVIEW

File: test/cpp/api/jit.cpp [hunk 1]
```diff
@@ -85,7 +85,16 @@ TEST(TorchScriptTest, TestNestedIValueModuleArgMatching) {
...
+}
 
 
-
+TEST(TorchScriptTest, TestDictArgMatching) {
+  auto module = torch::jit::compile(R"JIT(
+      def dict_op(a: Dict[str, Tensor], b: str):
+        return a[b]
+    )JIT");
+  c10::ivalue::DictUnorderedMap<torch::jit::IValue, torch::jit::IValue> dict;
```
Comment (eellison) [inline test/cpp/api/jit.cpp]: Can we add conversion for unordered_map<std::string, Tensor> ?

**Review by eellison — APPROVED** [review]
Looks good. In this PR or a follow up it might be nice to add conversion/specialization for unordered_map<std::string, Tensor>, and the other common cases.

File: test/cpp/api/jit.cpp (hunk 1 above)
Comment (driazati) [reply test/cpp/api/jit.cpp]: I'll leave that for a follow up since it involves a bunch changes to `IValue`

**Comment by sidazhang** [comment]
Can this diff support list[dict] in the C++ front end?

**Comment by driazati** [comment]
Any list of `IValue`s should be supported, so yes

**Pull request closed without merging**